### PR TITLE
TYP: __getitem__ method of EA (2nd pass)

### PR DESCRIPTION
**Pull request by simonjayhawkins**
follow-on from #37898

remove casts and ignores added in #37898 (needed to use typevar in DatetimeLikeArrayMixin in conjunction with overload for slice to return correct DatetimeArray/PeriodArray)

another pass required to

more typevars for return type of type(self)
type \_\_getitem__ in concrete classes (Categorical, IntervalArray and SparseArray)
overloads with Any report overlapping overloads
NAType resolves to `Any` since implemented in cython without stub file

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -484,6 +484,7 @@ def split_and_operate(
         -------
         list of blocks
         """
+        assert isinstance(self.values, np.ndarray)
```
Comment (simonjayhawkins) [inline pandas/core/internals/blocks.py:487]: need to ensure split_and_operate is not called from EA block (or base method not overridden)

File: pandas/core/indexes/period.py [hunk 2]
```diff
@@ -673,10 +673,7 @@ def difference(self, other, sort=None):
 
         if self.equals(other):
             # pass an empty PeriodArray with the appropriate dtype
-
-            # TODO: overload DatetimeLikeArrayMixin.__getitem__
-            values = cast(PeriodArray, self._data[:0])
-            return type(self)._simple_new(values, name=self.name)
+            return type(self)._simple_new(self._data[:0], name=self.name)
```
Comment (simonjayhawkins) [inline pandas/core/indexes/period.py:676]: reverting change from previous PR

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -4546,9 +4539,8 @@ def asof_locs(self, where: "Index", mask) -> np.ndarray:
 
         result = np.arange(len(self))[mask].take(locs)
 
-        # TODO: overload return type of ExtensionArray.__getitem__
-        first_value = cast(Any, self._values[mask.argmax()])
-        result[(locs == 0) & (where._values < first_value)] = -1
+        first = mask.argmax()
+        result[(locs == 0) & (where._values < self._values[first])] = -1
```
Comment (simonjayhawkins) [inline pandas/core/indexes/base.py:4543]: reverting change from previous PR

Comment (simonjayhawkins) [reply pandas/core/indexes/base.py:4543]: moved the assert to after the ndim check

File: pandas/core/internals/blocks.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/internals/blocks.py]: move assert after ndim check

**Comment by jreback** [comment]
lgtm. @jbrockmendel

File: pandas/core/arrays/_mixins.py [hunk 4]
```diff
@@ -214,9 +215,21 @@ def __setitem__(self, key, value):
     def _validate_setitem_value(self, value):
         return value
 
+    @overload
+    # error: Overloaded function signatures 1 and 2 overlap with incompatible
+    # return types  [misc]
+    def __getitem__(self, key: int) -> EAScalarOrMissing:  # type: ignore[misc]
```
Comment (jbrockmendel) [inline pandas/core/arrays/_mixins.py:221]: this can also return NDArrayBackedExtensionArrayT

File: pandas/core/arrays/datetimelike.py [hunk 5]
```diff
@@ -266,9 +267,19 @@ def __array__(self, dtype=None) -> np.ndarray:
             return np.array(list(self), dtype=object)
         return self._ndarray
 
+    @overload
+    def __getitem__(self, key: int) -> DTScalarOrNaT:
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:271]: same, can still return DatetimeLikeArrayT

**Comment by jbrockmendel** [comment]
couple of comments, otherwise LGTM

File: pandas/core/arrays/_mixins.py (hunk 4 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/_mixins.py:221]: can you elaborate. is this for 2d EA?

Comment (jbrockmendel) [reply pandas/core/arrays/_mixins.py:221]: NDArrayBackedExtensionArray supports 2D, exactly

Comment (simonjayhawkins) [reply pandas/core/arrays/_mixins.py:221]: so it also supports a tuple indexer? and I guess from that NDArrayBackedExtensionArray can't support nested data as it uses is_scalar checks.

so also need to change `EAScalarOrMissing = object  # both scalar value and na_value can be any type` -> `ScalarOrScalarMissing = Scalar  # both values and na_value must be scalars` ?

Comment (jbrockmendel) [reply pandas/core/arrays/_mixins.py:221]: > so it also supports a tuple indexer?

Yes.  For that matter i think even 1D will support 1-tuples

> and I guess from that NDArrayBackedExtensionArray can't support nested data as it uses is_scalar checks.

PandasArray can have object dtype, and Categorical can hold tuples.

**Comment by simonjayhawkins** [comment]
removing this from the 1.2 milestone for now.

since #35259 is merged, can now proceed with the outstanding tasks in parallel. so plan is to get all the tasks started (to get review) and will then circle back to this.

**Comment by simonjayhawkins** [comment]
closing while we discuss #39513

**Pull request closed without merging**